### ISSUE TITLE
The top level function ensureNotNull should be available not only for NullableEffect, EitherEffect and NullableEffect, but also for the Restrictable*Effect

### ISSUE THREAD

**Issue by lukaszgendek** [arrow-kt/arrow#2656]
Currently in the arrow.core 1.0.1 we are not able to use the ensureNotNull function in the eager scope.

For example the following code does not compile.

```
fun example(str: String?) : String? =
  nullable.eager {
    ensureNotNull(str)
  }

```
while this works fine:

```
suspend fun example(str: String?) : String? =
  nullable {
    ensureNotNull(str)
  }

```

This can be easily fixed by introducing the following function.

```
package arrow.core.computations

import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.contract

@OptIn(ExperimentalContracts::class) // Contracts not available on open functions, so made it top-level.
suspend fun <B : Any> RestrictedNullableEffect<*>.ensureNotNull(value: B?): B {
  contract {
    returns() implies (value != null)
  }
  return value.bind()
}

```
It would be nice to add the function to the arrow core library. The same applies to RestrictedOptionEffect and RestrictedEitherEffect.

**Comment by nomisRev** [comment]
Hey @lukaszgendek,

@i-walker is looking into moving https://github.com/nomisRev/Continuation to Arrow to improve the runtime and solve this outstanding issues. With this encoding we can define it once for all types, similar to what you proposed in the ticket.

**Comment by i-walker** [comment]
https://github.com/arrow-kt/arrow/pull/2660

**Comment by i-walker** [comment]
Continued in #2661

**Comment by i-walker** [comment]
Resolved in #2661
